lib/std/time.zig

```diff
@@ -89,202 +119,297 @@ pub fn nanoTimestamp() i128 {
-            os.darwin.mach_timebase_info(&freq);
-
-            return Timer{
-                .frequency = freq,
-                .resolution = @divFloor(freq.numer, freq.denom),
-                .start_time = os.darwin.mach_absolute_time(),
-            };
-        } else {
-            // On Linux, seccomp can do arbitrary things to our ability to call
-            // syscalls, including return any errno value it wants and
-            // inconsistently throwing errors. Since we can't account for
-            // abuses of seccomp in a reasonable way, we'll assume that if
-            // seccomp is going to block us it will at least do so consistently
-            var res: os.timespec = undefined;
-            os.clock_getres(monotonic_clock_id, &res) catch return error.TimerUnsupported;
-
-            var ts: os.timespec = undefined;
-            os.clock_gettime(monotonic_clock_id, &ts) catch return error.TimerUnsupported;
-
-            return Timer{
-                .resolution = @intCast(u64, res.tv_sec) * ns_per_s + @intCast(u64, res.tv_nsec),
-                .start_time = @intCast(u64, ts.tv_sec) * ns_per_s + @intCast(u64, ts.tv_nsec),
-                .frequency = {},
-            };
+    const Static = struct {
+        var last_now: u64 = 0;
+        var last_now_mutex = std.Thread.Mutex{};
+    };
+
+    // If 64bit atomics are available, its generally cheaper to use them vs locking below.
+    if (comptime supports64bitAtomics()) {
+        var last = atomic.load(&Static.last_now, .Relaxed);
+        while (true) {
+            if (last >= current_now) {
+                return last;
+            }
+            last = atomic.tryCompareAndSwap(
+                &Static.last_now,
+                last,
+                current_now,
+                .Relaxed,
+                .Relaxed,
+            ) orelse return current_now;
         }
+    }
+
+    // For platforms that don't support 64bit atomics, fall back to Lock based synchronization.
+    // We use a Lock over a Mutex here as the former is unfair and may deal with micro-contention better.
+    const held = Static.last_now_mutex.acquire();
+    defer held.release();
 
-        return self;
+    const last_now = Static.last_now;
+    if (last_now > current_now) {
+        return last_now;
     }
 
-    /// Reads the timer value since start or the last reset in nanoseconds
-    pub fn read(self: Timer) u64 {
-        var clock = clockNative() - self.start_time;
-        return self.nativeDurationToNanos(clock);
+    Static.last_now = current_now;
+    return current_now;
+}
+
+/// Returns true if the platform supports 64-bit atomic operations.
+fn supports64bitAtomics() bool {
+    // TODO: check if this is actually true.
+    return @sizeOf(usize) >= @sizeOf(u64);
+}
+
+/// Returns true if the system's high performance timer implementation is actually monotonic.
+/// For some platforms, this is not always the case: https://doc.rust-lang.org/src/std/time.rs.html#227
```

Comment:
```suggestion
/// For some platforms, this is not always the case: https://doc.rust-lang.org/1.50.0/src/std/time.rs.html#227
```

Reply:
future versions of rust may move that file/code, so it is better(IMO) to link to a permlink